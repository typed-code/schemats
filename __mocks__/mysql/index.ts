const mysqlStub = {
  query: jest.fn(),
};

export = {
  createConnection: (connection: string) =>
    connection
      ? mysqlStub
      : {
          mysqlStub: {
            ...mysqlStub,
            withResults: (results: any[]) =>
              mysqlStub.query.mockImplementation((query: string, _: any, cb: any) =>
                cb(null, results)
              ),
            withError: (error: string) =>
              mysqlStub.query.mockImplementation((query: string, _: any, cb: any) => cb(error)),
          },
          resetMocks: () => {
            mysqlStub.query.mockReset();
          },
        },
};
