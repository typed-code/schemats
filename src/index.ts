/**
 * Schemats takes sql database schema and creates corresponding typescript definitions
 * Created by xiamx on 2016-08-10.
 */

import * as pkgUp from 'pkg-up';
import * as prettier from 'prettier';
import Options, { OptionValues } from './options';
import { Database, getDatabase } from './schema';
import { generateEnumType, generateTableInterface, generateTableTypes } from './typescript';

const pkgPath = pkgUp.sync();
const pkgVersion = pkgPath ? require(pkgPath).version : 'UNKNOWN';

function getTime() {
  const padTime = (value: number) => `0${value}`.slice(-2);
  const time = new Date();
  const yyyy = time.getFullYear();
  const MM = padTime(time.getMonth() + 1);
  const dd = padTime(time.getDate());
  const hh = padTime(time.getHours());
  const mm = padTime(time.getMinutes());
  const ss = padTime(time.getSeconds());
  return `${yyyy}-${MM}-${dd} ${hh}:${mm}:${ss}`;
}

function buildHeader(
  db: Database,
  tables: string[],
  schema: string | null,
  options: OptionValues
): string {
  const commands = [
    'schemats',
    'generate',
    '-c',
    db.connectionString.replace(/:\/\/.*@/, '://username:password@'),
  ];
  if (options.camelCase) {
    commands.push('-C');
  }
  if (tables.length > 0) {
    tables.forEach((t: string) => {
      commands.push('-t', t);
    });
  }
  if (schema) {
    commands.push('-s', schema);
  }

  return `
        /**
         * AUTO-GENERATED FILE @ ${getTime()} - DO NOT EDIT!
         *
         * This file was automatically generated by schemats v.${pkgVersion}
         * $ ${commands.join(' ')}
         *
         */

    `;
}

export async function typescriptOfTable(
  db: Database | string,
  table: string,
  schema: string,
  options = new Options()
) {
  if (typeof db === 'string') {
    db = getDatabase(db);
  }

  const tableTypes = await db.getTableTypes(table, schema, options);

  let interfaces = '';
  interfaces += generateTableTypes(table, tableTypes, options);
  interfaces += generateTableInterface(table, tableTypes, options);
  return interfaces;
}

export async function typescriptOfSchema(
  db: Database | string,
  tables: string[] = [],
  schema: string | null = null,
  options: OptionValues = {}
): Promise<string> {
  if (typeof db === 'string') {
    db = getDatabase(db);
  }

  if (!schema) {
    schema = db.getDefaultSchema();
  }

  if (tables.length === 0) {
    tables = await db.getSchemaTables(schema);
  }

  const optionsObject = new Options(options);

  const enumTypes = generateEnumType(await db.getEnumTypes(schema), optionsObject);
  const interfacePromises = tables.map(table =>
    typescriptOfTable(db, table, schema as string, optionsObject)
  );
  const interfaces = await Promise.all(interfacePromises).then(tsOfTable => tsOfTable.join(''));

  let output = '/* tslint:disable */\n\n';
  if (optionsObject.options.writeHeader) {
    output += buildHeader(db, tables, schema, options);
  }
  output += enumTypes;
  output += interfaces;

  const configFilePath = await prettier.resolveConfigFile();
  const prettierOptions = await prettier.resolveConfig(configFilePath || '');
  return prettier.format(output, { ...prettierOptions, parser: 'typescript' });
}

export { Database, getDatabase } from './schema';
export { Options, OptionValues };
