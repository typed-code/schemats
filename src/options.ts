import { camelCase, upperFirst } from 'lodash';

const DEFAULT_OPTIONS: OptionValues = {
  writeHeader: true,
  camelCase: false,
};

export interface OptionValues {
  camelCase?: boolean;
  writeHeader?: boolean; // write schemats description header
}

export default class Options {
  public options: OptionValues;

  constructor(options: OptionValues = {}) {
    this.options = { ...DEFAULT_OPTIONS, ...options };
  }

  public transformTypeName(typename: string) {
    return this.options.camelCase ? upperFirst(camelCase(typename)) : typename;
  }

  public transformColumnName(columnName: string) {
    return this.options.camelCase ? camelCase(columnName) : columnName;
  }
}
