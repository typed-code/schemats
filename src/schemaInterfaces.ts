import Options from './options';

export interface ColumnDefinition {
  udtName: string;
  nullable: boolean;
  tsType?: string;
}

export interface TableDefinition {
  [columnName: string]: ColumnDefinition;
}

export interface Database {
  connectionString: string;

  query(queryString: string): Promise<object[]>;

  getDefaultSchema(): string;

  getEnumTypes(schema?: string): Promise<{ [key: string]: string[] }>;

  getTableDefinition(tableName: string, tableSchema: string): Promise<TableDefinition>;

  getTableTypes(tableName: string, tableSchema: string, options: Options): Promise<TableDefinition>;

  getSchemaTables(schemaName: string): Promise<string[]>;
}
