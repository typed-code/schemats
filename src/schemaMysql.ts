import { isEqual, keys, mapValues } from 'lodash';
import { Connection, createConnection, MysqlError } from 'mysql';
import { parse as urlParse } from 'url';
import { Options } from './options';
import { Database, ICustomTypes, ITable } from './schemaInterfaces';

export class MysqlDatabase implements Database {
  private db: Connection;
  private readonly defaultSchema: string;

  constructor(public connectionString: string) {
    this.db = createConnection(connectionString);
    const url = urlParse(connectionString, true);
    this.defaultSchema = url && url.pathname ? url.pathname.substr(1) : 'public';
  }

  // uses the type mappings from https://github.com/mysqljs/ where sensible
  private static mapTableDefinitionToType(
    tableDefinition: ITable,
    customTypes: string[],
    options: Options
  ): ITable {
    if (!options) {
      throw new Error('No options given');
    }

    tableDefinition.columns = mapValues(tableDefinition.columns, column => {
      switch (column.udtName) {
        case 'char':
        case 'varchar':
        case 'text':
        case 'tinytext':
        case 'mediumtext':
        case 'longtext':
        case 'time':
        case 'geometry':
        case 'set':
        case 'enum':
          // keep set and enum defaulted to string if custom type not mapped
          column.tsType = 'string';
          return column;
        case 'integer':
        case 'int':
        case 'smallint':
        case 'mediumint':
        case 'bigint':
        case 'double':
        case 'decimal':
        case 'numeric':
        case 'float':
        case 'year':
          column.tsType = 'number';
          return column;
        case 'tinyint':
          column.tsType = 'boolean';
          return column;
        case 'json':
          column.tsType = 'Object';
          return column;
        case 'date':
        case 'datetime':
        case 'timestamp':
          column.tsType = 'Date';
          return column;
        case 'tinyblob':
        case 'mediumblob':
        case 'longblob':
        case 'blob':
        case 'binary':
        case 'varbinary':
        case 'bit':
          column.tsType = 'Buffer';
          return column;
        default:
          if (customTypes.includes(column.udtName)) {
            column.tsType = options.transformTypeName(column.udtName);
            return column;
          } else {
            // tslint:disable-next-line
            console.log(
              `Type '${column.udtName}' has been mapped to 'any' because no specific type has been found.`
            );
            column.tsType = 'any';
            return column;
          }
      }
    });

    return tableDefinition;
  }

  private static parseMysqlEnumeration(mysqlEnum: string): string[] {
    return mysqlEnum
      .replace(/(^(enum|set)\('|'\)$)/gi, '')
      .split(`','`)
      .sort();
  }

  private static getEnumNameFromColumn(dataType: string, columnName: string): string {
    return `${columnName}_${dataType}`;
  }

  public query<T = object>(queryString: string) {
    return this.queryAsync<T>(queryString);
  }

  public async getEnumTypes(schema?: string, tables: string[] = []): Promise<ICustomTypes> {
    let additionWhereClause: string = '';
    const params: string[] = [];
    if (schema) {
      additionWhereClause = `and table_schema = ? `;
      params.push(schema);
    }

    if (tables.length > 0) {
      additionWhereClause = `and table_name in (${tables.map(_ => '?').join(',')}) `;
      params.push(...tables);
    }

    const rawEnumRecords = await this.queryAsync<{
      table_name: string;
      column_name: string;
      column_type: string;
      data_type: string;
    }>(
      'SELECT table_name, column_name, column_type, data_type ' +
        'FROM information_schema.columns ' +
        `WHERE data_type IN ('enum', 'set') ${additionWhereClause}ORDER BY table_name, column_name`,
      params
    );

    const groupedEnums = rawEnumRecords.reduce(
      (enums, { table_name, column_name, data_type, column_type }) => {
        const enumName = MysqlDatabase.getEnumNameFromColumn(data_type, column_name);
        const enumValues = MysqlDatabase.parseMysqlEnumeration(column_type);

        enums[enumName] = enums[enumName] || {};
        enums[enumName][table_name] = enumValues;

        return enums;
      },
      ({} as unknown) as { [enumName: string]: { [tableName: string]: string[] } }
    );

    return Object.entries(groupedEnums).reduce((result, [enumName, tablesMap]) => {
      const tableKeys = Object.keys(tablesMap);
      const firstTableValues = tablesMap[tableKeys[0]];
      const allValuesSame = tableKeys.every(tableKey =>
        isEqual(tablesMap[tableKey], firstTableValues)
      );

      if (tableKeys.length === 1 || allValuesSame) {
        result[enumName] = tablesMap[tableKeys[0]];
      } else {
        tableKeys.forEach(tableKey => {
          result[`${tableKey}_${enumName}`] = tablesMap[tableKey];
        });
      }

      return result;
    }, ({} as unknown) as ICustomTypes);
  }

  public async getTablesDefinition(
    tableNames: string[],
    tableSchema: string,
    customTypes: ICustomTypes
  ): Promise<ITable[]> {
    const params = [tableSchema];
    let whereClauseAddition = '';

    if (tableNames.length) {
      params.push(...tableNames);
      whereClauseAddition = `and table_name IN (${tableNames.map(_ => '?').join(',')}) `;
    }

    const tableColumns = await this.queryAsync<{
      table_name: string;
      column_name: string;
      data_type: string;
      is_nullable: 'YES' | 'NO';
    }>(
      'SELECT table_name, column_name, data_type, is_nullable ' +
        'FROM information_schema.columns ' +
        'WHERE table_schema = ? ' +
        whereClauseAddition +
        'ORDER BY table_name, column_name',
      params
    );

    const tablesMap = tableColumns.reduce(
      (result, { table_name, column_name, data_type, is_nullable }) => {
        result[table_name] = result[table_name] || {
          name: table_name,
          columns: {},
        };

        const table = result[table_name];

        const enumName = MysqlDatabase.getEnumNameFromColumn(data_type, column_name);

        table.columns[column_name] = {
          udtName: /^(enum|set)$/i.test(data_type)
            ? customTypes.hasOwnProperty(enumName)
              ? enumName
              : `${table_name}_${enumName}`
            : data_type,
          nullable: is_nullable === 'YES',
          tsType: '',
        };

        return result;
      },
      ({} as unknown) as { [tableName: string]: ITable }
    );

    return Object.values(tablesMap);
  }

  public async getTablesTypes(
    tableNames: string[],
    tableSchema: string,
    customTypes: ICustomTypes,
    options: Options
  ): Promise<ITable[]> {
    const customTypesKeys = keys(customTypes);

    const tableDefinitions = await this.getTablesDefinition(tableNames, tableSchema, customTypes);

    return tableDefinitions.map(table =>
      MysqlDatabase.mapTableDefinitionToType(table, customTypesKeys, options)
    );
  }

  public async getSchemaTables(schemaName: string): Promise<string[]> {
    const schemaTables = await this.queryAsync<{ table_name: string }>(
      'SELECT table_name ' +
        'FROM information_schema.columns ' +
        'WHERE table_schema = ? ' +
        'GROUP BY table_name ORDER BY table_name',
      [schemaName]
    );
    return schemaTables.map(({ table_name }) => table_name);
  }

  public getDefaultSchema(): string {
    return this.defaultSchema;
  }

  private queryAsync<T = object>(queryString: string, escapedValues?: string[]): Promise<T[]> {
    return new Promise((resolve, reject) => {
      this.db.query(queryString, escapedValues, (error: MysqlError, results: T[]) => {
        if (error) {
          return reject(error);
        }
        return resolve(this.toLowerCaseColumnName<T>(results));
      });
    });
  }

  private toLowerCaseColumnName<T>(results: T[]): T[] {
    return results.map((row: any) =>
      Object.keys(row).reduce((newRow, key) => {
        newRow[key.toLowerCase()] = row[key];
        return newRow;
      }, {} as any)
    );
  }
}
