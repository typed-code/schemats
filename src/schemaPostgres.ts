import { keys, mapValues } from 'lodash';
import * as PgPromise from 'pg-promise';
import { Options } from './options';

import { Database, ITable } from './schemaInterfaces';

const pgp = PgPromise();

export class PostgresDatabase implements Database {
  private db: PgPromise.IDatabase<{}>;

  constructor(public connectionString: string) {
    this.db = pgp(connectionString);
  }

  private static mapTableDefinitionToType(
    tableDefinition: ITable,
    customTypes: string[],
    options: Options
  ): ITable {
    tableDefinition.columns = mapValues(tableDefinition.columns, column => {
      switch (column.udtName) {
        case 'bpchar':
        case 'char':
        case 'varchar':
        case 'text':
        case 'citext':
        case 'uuid':
        case 'bytea':
        case 'inet':
        case 'time':
        case 'timetz':
        case 'interval':
        case 'name':
          column.tsType = 'string';
          return column;
        case 'int2':
        case 'int4':
        case 'int8':
        case 'float4':
        case 'float8':
        case 'numeric':
        case 'money':
        case 'oid':
          column.tsType = 'number';
          return column;
        case 'bool':
          column.tsType = 'boolean';
          return column;
        case 'json':
        case 'jsonb':
          column.tsType = 'Object';
          return column;
        case 'date':
        case 'timestamp':
        case 'timestamptz':
          column.tsType = 'Date';
          return column;
        case '_int2':
        case '_int4':
        case '_int8':
        case '_float4':
        case '_float8':
        case '_numeric':
        case '_money':
          column.tsType = 'Array<number>';
          return column;
        case '_bool':
          column.tsType = 'Array<boolean>';
          return column;
        case '_varchar':
        case '_text':
        case '_citext':
        case '_uuid':
        case '_bytea':
          column.tsType = 'Array<string>';
          return column;
        case '_json':
        case '_jsonb':
          column.tsType = 'Array<Object>';
          return column;
        case '_timestamptz':
          column.tsType = 'Array<Date>';
          return column;
        default:
          if (customTypes.indexOf(column.udtName) !== -1) {
            column.tsType = options.transformTypeName(column.udtName);
            return column;
          } else {
            // tslint:disable-next-line
            console.log(
              `Type '${column.udtName}' has been mapped to 'any' because no specific type has been found.`
            );
            column.tsType = 'any';
            return column;
          }
      }
    });
    return tableDefinition;
  }

  public query(queryString: string) {
    return this.db.query(queryString);
  }

  public async getEnumTypes(
    schema?: string,
    tables: string[] = []
  ): Promise<{ [key: string]: string[] }> {
    interface T {
      name: string;
      value: any;
    }

    const enums: any = {};
    const enumSchemaWhereClause = schema ? pgp.as.format(`where n.nspname = $1`, schema) : '';
    await this.db.each<T>(
      'select n.nspname as schema, t.typname as name, e.enumlabel as value ' +
        'from pg_type t ' +
        'join pg_enum e on t.oid = e.enumtypid ' +
        'join pg_catalog.pg_namespace n ON n.oid = t.typnamespace ' +
        `${enumSchemaWhereClause} ` +
        'order by t.typname asc, e.enumlabel asc;',
      [],
      (item: T) => {
        if (!enums[item.name]) {
          enums[item.name] = [];
        }
        enums[item.name].push(item.value);
      }
    );
    return enums;
  }

  public async getTablesDefinition(tableNames: string[], tableSchema: string): Promise<ITable[]> {
    interface T {
      table_name: string;
      column_name: string;
      udt_name: string;
      is_nullable: string;
    }

    const tablesMap: { [tableName: string]: ITable } = {};
    let whereClauseAddition = '';
    const params: any[] = [tableSchema];

    if (tableNames.length) {
      whereClauseAddition = `and table_name IN ($2:csv) `;
      params.push(tableNames);
    }

    await this.db.each<T>(
      'SELECT table_name, column_name, udt_name, is_nullable ' +
        'FROM information_schema.columns ' +
        `WHERE table_schema = $1 ${whereClauseAddition}` +
        'ORDER BY table_name, column_name',
      params,
      (schemaItem: T) => {
        tablesMap[schemaItem.table_name] = tablesMap[schemaItem.table_name] || {
          name: schemaItem.table_name,
          columns: {},
        };
        const table = tablesMap[schemaItem.table_name];

        table.columns[schemaItem.column_name] = {
          udtName: schemaItem.udt_name,
          nullable: schemaItem.is_nullable === 'YES',
          tsType: '',
        };
      }
    );

    return Object.values(tablesMap);
  }

  public async getTablesTypes(
    tableNames: string[],
    tableSchema: string,
    options: Options
  ): Promise<ITable[]> {
    const enumTypes = await this.getEnumTypes();
    const customTypes = keys(enumTypes);
    const tableDefinitions = await this.getTablesDefinition(tableNames, tableSchema);

    return tableDefinitions.map(table =>
      PostgresDatabase.mapTableDefinitionToType(table, customTypes, options)
    );
  }

  public async getSchemaTables(schemaName: string): Promise<string[]> {
    return this.db.map<string>(
      'SELECT table_name ' +
        'FROM information_schema.columns ' +
        'WHERE table_schema = $1 ' +
        'GROUP BY table_name ORDER BY table_name',
      [schemaName],
      (schemaItem: { table_name: string }) => schemaItem.table_name
    );
  }

  public getDefaultSchema(): string {
    return 'public';
  }
}
