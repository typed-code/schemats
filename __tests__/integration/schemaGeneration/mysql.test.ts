import * as mysql from 'mysql';
import { Database, getDatabase } from '../../../src';
import { condBeforeAll, condDescribe, generateTypes, loadSchema } from '../../testUtility';

describe('schema generation integration testing', () => {
  condDescribe(!!process.env.MYSQL_URL, this, 'mysql', () => {
    let db: Database;

    condBeforeAll(!!process.env.MYSQL_URL, this, async () => {
      (mysql as any).disableStub();
      db = getDatabase(`${process.env.MYSQL_URL}?multipleStatements=true`);
      await loadSchema(db, './__tests__/fixture/mysql/initCleanup.sql');
    });

    afterAll(() => {
      (mysql as any).enableStub();
    });

    it('Basic generation', async () => {
      const inputSQLFile = '__tests__/fixture/mysql/osm.sql';
      const output = await generateTypes(
        inputSQLFile,
        {
          tables: ['users', 'user_enums', 'package'],
        },
        db
      );

      expect(output).toMatchSnapshot();
    });

    it('Camelcase generation', async () => {
      const inputSQLFile = '__tests__/fixture/mysql/osm.sql';
      const output = await generateTypes(
        inputSQLFile,
        {
          tables: ['users', 'user_enums', 'package'],
          camelCase: true,
        },
        db
      );

      expect(output).toMatchSnapshot();
    });

    it('Enum conflict in columns', async () => {
      const inputSQLFile = '__tests__/fixture/mysql/conflict.sql';
      try {
        await generateTypes(
          inputSQLFile,
          {
            tables: ['location', 'location_history'],
            writeHeader: false,
          },
          db
        );
      } catch (e) {
        expect(e.message).toEqual(
          'Multiple enums with the same name and contradicting types were found: location_type: ["city","province","country"] and ["city","state","country"]'
        );
      }
    });
  });
});
