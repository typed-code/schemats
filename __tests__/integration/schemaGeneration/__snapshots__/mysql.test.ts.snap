// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema generation integration testing mysql Basic generation 1`] = `
"/* tslint:disable */

export type enum_col_enum = 'enum1' | 'enum2' | 'enum3';
export type nullable_enum_col_enum = 'enum1' | 'enum2' | 'enum3';
export type nullable_set_col_set = 'set1' | 'set2' | 'set3';
export type set_col_set = 'set1' | 'set2' | 'set3';

export namespace packageFields {
  export type number_ = number;
  export type string_ = string;
}

export interface package_ {
  number: packageFields.number_;
  string: packageFields.string_;
}

export namespace user_enumsFields {
  export type enum_col = enum_col_enum;
  export type nullable_enum_col = nullable_enum_col_enum | null;
  export type nullable_set_col = nullable_set_col_set | null;
  export type set_col = set_col_set;
}

export interface user_enums {
  enum_col: user_enumsFields.enum_col;
  nullable_enum_col: user_enumsFields.nullable_enum_col;
  nullable_set_col: user_enumsFields.nullable_set_col;
  set_col: user_enumsFields.set_col;
}

export namespace usersFields {
  export type bigint_col = number;
  export type binary_col = Buffer;
  export type bit_col = Buffer;
  export type blob_col = Buffer;
  export type char_col = string;
  export type date_col = Date;
  export type datetime_col = Date;
  export type decimal_col = number;
  export type double_col = number;
  export type enum_col = enum_col_enum;
  export type float_col = number;
  export type geometry_col = string;
  export type int_col = number;
  export type integer_col = number;
  export type json_col = Object;
  export type longblob_col = Buffer;
  export type longtext_col = string;
  export type mediumblob_col = Buffer;
  export type mediumint_col = number;
  export type mediumtext_col = string;
  export type nullable_bigint_col = number | null;
  export type nullable_binary_col = Buffer | null;
  export type nullable_bit_col = Buffer | null;
  export type nullable_blob_col = Buffer | null;
  export type nullable_char_col = string | null;
  export type nullable_date_col = Date | null;
  export type nullable_datetime_col = Date | null;
  export type nullable_decimal_col = number | null;
  export type nullable_double_col = number | null;
  export type nullable_enum_col = nullable_enum_col_enum | null;
  export type nullable_float_col = number | null;
  export type nullable_geometry_col = string | null;
  export type nullable_int_col = number | null;
  export type nullable_integer_col = number | null;
  export type nullable_json_col = Object | null;
  export type nullable_longblob_col = Buffer | null;
  export type nullable_longtext_col = string | null;
  export type nullable_mediumblob_col = Buffer | null;
  export type nullable_mediumint_col = number | null;
  export type nullable_mediumtext_col = string | null;
  export type nullable_numeric_col = number | null;
  export type nullable_set_col = nullable_set_col_set | null;
  export type nullable_smallint_col = number | null;
  export type nullable_text_col = string | null;
  export type nullable_time_col = string | null;
  export type nullable_tinyblob_col = Buffer | null;
  export type nullable_tinyint_col = boolean | null;
  export type nullable_tinytext_col = string | null;
  export type nullable_varbinary_col = Buffer | null;
  export type nullable_year_col = number | null;
  export type numeric_col = number;
  export type set_col = set_col_set;
  export type smallint_col = number;
  export type text_col = string;
  export type time_col = string;
  export type timestamp_col = Date;
  export type tinyblob_col = Buffer;
  export type tinyint_col = boolean;
  export type tinytext_col = string;
  export type varbinary_col = Buffer;
  export type year_col = number;
}

export interface users {
  bigint_col: usersFields.bigint_col;
  binary_col: usersFields.binary_col;
  bit_col: usersFields.bit_col;
  blob_col: usersFields.blob_col;
  char_col: usersFields.char_col;
  date_col: usersFields.date_col;
  datetime_col: usersFields.datetime_col;
  decimal_col: usersFields.decimal_col;
  double_col: usersFields.double_col;
  enum_col: usersFields.enum_col;
  float_col: usersFields.float_col;
  geometry_col: usersFields.geometry_col;
  int_col: usersFields.int_col;
  integer_col: usersFields.integer_col;
  json_col: usersFields.json_col;
  longblob_col: usersFields.longblob_col;
  longtext_col: usersFields.longtext_col;
  mediumblob_col: usersFields.mediumblob_col;
  mediumint_col: usersFields.mediumint_col;
  mediumtext_col: usersFields.mediumtext_col;
  nullable_bigint_col: usersFields.nullable_bigint_col;
  nullable_binary_col: usersFields.nullable_binary_col;
  nullable_bit_col: usersFields.nullable_bit_col;
  nullable_blob_col: usersFields.nullable_blob_col;
  nullable_char_col: usersFields.nullable_char_col;
  nullable_date_col: usersFields.nullable_date_col;
  nullable_datetime_col: usersFields.nullable_datetime_col;
  nullable_decimal_col: usersFields.nullable_decimal_col;
  nullable_double_col: usersFields.nullable_double_col;
  nullable_enum_col: usersFields.nullable_enum_col;
  nullable_float_col: usersFields.nullable_float_col;
  nullable_geometry_col: usersFields.nullable_geometry_col;
  nullable_int_col: usersFields.nullable_int_col;
  nullable_integer_col: usersFields.nullable_integer_col;
  nullable_json_col: usersFields.nullable_json_col;
  nullable_longblob_col: usersFields.nullable_longblob_col;
  nullable_longtext_col: usersFields.nullable_longtext_col;
  nullable_mediumblob_col: usersFields.nullable_mediumblob_col;
  nullable_mediumint_col: usersFields.nullable_mediumint_col;
  nullable_mediumtext_col: usersFields.nullable_mediumtext_col;
  nullable_numeric_col: usersFields.nullable_numeric_col;
  nullable_set_col: usersFields.nullable_set_col;
  nullable_smallint_col: usersFields.nullable_smallint_col;
  nullable_text_col: usersFields.nullable_text_col;
  nullable_time_col: usersFields.nullable_time_col;
  nullable_tinyblob_col: usersFields.nullable_tinyblob_col;
  nullable_tinyint_col: usersFields.nullable_tinyint_col;
  nullable_tinytext_col: usersFields.nullable_tinytext_col;
  nullable_varbinary_col: usersFields.nullable_varbinary_col;
  nullable_year_col: usersFields.nullable_year_col;
  numeric_col: usersFields.numeric_col;
  set_col: usersFields.set_col;
  smallint_col: usersFields.smallint_col;
  text_col: usersFields.text_col;
  time_col: usersFields.time_col;
  timestamp_col: usersFields.timestamp_col;
  tinyblob_col: usersFields.tinyblob_col;
  tinyint_col: usersFields.tinyint_col;
  tinytext_col: usersFields.tinytext_col;
  varbinary_col: usersFields.varbinary_col;
  year_col: usersFields.year_col;
}
"
`;

exports[`schema generation integration testing mysql Camelcase generation 1`] = `
"/* tslint:disable */

export type EnumColEnum = 'enum1' | 'enum2' | 'enum3';
export type NullableEnumColEnum = 'enum1' | 'enum2' | 'enum3';
export type NullableSetColSet = 'set1' | 'set2' | 'set3';
export type SetColSet = 'set1' | 'set2' | 'set3';

export namespace PackageFields {
  export type number_ = number;
  export type string_ = string;
}

export interface Package {
  number: PackageFields.number_;
  string: PackageFields.string_;
}

export namespace UserEnumsFields {
  export type enumCol = EnumColEnum;
  export type nullableEnumCol = NullableEnumColEnum | null;
  export type nullableSetCol = NullableSetColSet | null;
  export type setCol = SetColSet;
}

export interface UserEnums {
  enumCol: UserEnumsFields.enumCol;
  nullableEnumCol: UserEnumsFields.nullableEnumCol;
  nullableSetCol: UserEnumsFields.nullableSetCol;
  setCol: UserEnumsFields.setCol;
}

export namespace UsersFields {
  export type bigintCol = number;
  export type binaryCol = Buffer;
  export type bitCol = Buffer;
  export type blobCol = Buffer;
  export type charCol = string;
  export type dateCol = Date;
  export type datetimeCol = Date;
  export type decimalCol = number;
  export type doubleCol = number;
  export type enumCol = EnumColEnum;
  export type floatCol = number;
  export type geometryCol = string;
  export type intCol = number;
  export type integerCol = number;
  export type jsonCol = Object;
  export type longblobCol = Buffer;
  export type longtextCol = string;
  export type mediumblobCol = Buffer;
  export type mediumintCol = number;
  export type mediumtextCol = string;
  export type nullableBigintCol = number | null;
  export type nullableBinaryCol = Buffer | null;
  export type nullableBitCol = Buffer | null;
  export type nullableBlobCol = Buffer | null;
  export type nullableCharCol = string | null;
  export type nullableDateCol = Date | null;
  export type nullableDatetimeCol = Date | null;
  export type nullableDecimalCol = number | null;
  export type nullableDoubleCol = number | null;
  export type nullableEnumCol = NullableEnumColEnum | null;
  export type nullableFloatCol = number | null;
  export type nullableGeometryCol = string | null;
  export type nullableIntCol = number | null;
  export type nullableIntegerCol = number | null;
  export type nullableJsonCol = Object | null;
  export type nullableLongblobCol = Buffer | null;
  export type nullableLongtextCol = string | null;
  export type nullableMediumblobCol = Buffer | null;
  export type nullableMediumintCol = number | null;
  export type nullableMediumtextCol = string | null;
  export type nullableNumericCol = number | null;
  export type nullableSetCol = NullableSetColSet | null;
  export type nullableSmallintCol = number | null;
  export type nullableTextCol = string | null;
  export type nullableTimeCol = string | null;
  export type nullableTinyblobCol = Buffer | null;
  export type nullableTinyintCol = boolean | null;
  export type nullableTinytextCol = string | null;
  export type nullableVarbinaryCol = Buffer | null;
  export type nullableYearCol = number | null;
  export type numericCol = number;
  export type setCol = SetColSet;
  export type smallintCol = number;
  export type textCol = string;
  export type timeCol = string;
  export type timestampCol = Date;
  export type tinyblobCol = Buffer;
  export type tinyintCol = boolean;
  export type tinytextCol = string;
  export type varbinaryCol = Buffer;
  export type yearCol = number;
}

export interface Users {
  bigintCol: UsersFields.bigintCol;
  binaryCol: UsersFields.binaryCol;
  bitCol: UsersFields.bitCol;
  blobCol: UsersFields.blobCol;
  charCol: UsersFields.charCol;
  dateCol: UsersFields.dateCol;
  datetimeCol: UsersFields.datetimeCol;
  decimalCol: UsersFields.decimalCol;
  doubleCol: UsersFields.doubleCol;
  enumCol: UsersFields.enumCol;
  floatCol: UsersFields.floatCol;
  geometryCol: UsersFields.geometryCol;
  intCol: UsersFields.intCol;
  integerCol: UsersFields.integerCol;
  jsonCol: UsersFields.jsonCol;
  longblobCol: UsersFields.longblobCol;
  longtextCol: UsersFields.longtextCol;
  mediumblobCol: UsersFields.mediumblobCol;
  mediumintCol: UsersFields.mediumintCol;
  mediumtextCol: UsersFields.mediumtextCol;
  nullableBigintCol: UsersFields.nullableBigintCol;
  nullableBinaryCol: UsersFields.nullableBinaryCol;
  nullableBitCol: UsersFields.nullableBitCol;
  nullableBlobCol: UsersFields.nullableBlobCol;
  nullableCharCol: UsersFields.nullableCharCol;
  nullableDateCol: UsersFields.nullableDateCol;
  nullableDatetimeCol: UsersFields.nullableDatetimeCol;
  nullableDecimalCol: UsersFields.nullableDecimalCol;
  nullableDoubleCol: UsersFields.nullableDoubleCol;
  nullableEnumCol: UsersFields.nullableEnumCol;
  nullableFloatCol: UsersFields.nullableFloatCol;
  nullableGeometryCol: UsersFields.nullableGeometryCol;
  nullableIntCol: UsersFields.nullableIntCol;
  nullableIntegerCol: UsersFields.nullableIntegerCol;
  nullableJsonCol: UsersFields.nullableJsonCol;
  nullableLongblobCol: UsersFields.nullableLongblobCol;
  nullableLongtextCol: UsersFields.nullableLongtextCol;
  nullableMediumblobCol: UsersFields.nullableMediumblobCol;
  nullableMediumintCol: UsersFields.nullableMediumintCol;
  nullableMediumtextCol: UsersFields.nullableMediumtextCol;
  nullableNumericCol: UsersFields.nullableNumericCol;
  nullableSetCol: UsersFields.nullableSetCol;
  nullableSmallintCol: UsersFields.nullableSmallintCol;
  nullableTextCol: UsersFields.nullableTextCol;
  nullableTimeCol: UsersFields.nullableTimeCol;
  nullableTinyblobCol: UsersFields.nullableTinyblobCol;
  nullableTinyintCol: UsersFields.nullableTinyintCol;
  nullableTinytextCol: UsersFields.nullableTinytextCol;
  nullableVarbinaryCol: UsersFields.nullableVarbinaryCol;
  nullableYearCol: UsersFields.nullableYearCol;
  numericCol: UsersFields.numericCol;
  setCol: UsersFields.setCol;
  smallintCol: UsersFields.smallintCol;
  textCol: UsersFields.textCol;
  timeCol: UsersFields.timeCol;
  timestampCol: UsersFields.timestampCol;
  tinyblobCol: UsersFields.tinyblobCol;
  tinyintCol: UsersFields.tinyintCol;
  tinytextCol: UsersFields.tinytextCol;
  varbinaryCol: UsersFields.varbinaryCol;
  yearCol: UsersFields.yearCol;
}
"
`;
